### ISSUE TITLE
FR: Get size of current batch

### ISSUE THREAD

**Issue by jaschaio** [firebase/firebase-js-sdk#1560]
* Operating System version: Node v8.11
  * Firebase SDK version: 5.8.4
  * Firebase Product: Firestore

Firestore Batched Writes are limited to 500 documents per batch. It would be handy to have a method on the batch itself that returns the number of currently queued operations so that we can do a commit when needed.

It is actually already possible using the internal `_writes` array length, but it doesn't feel right:

```javascript
// … Define collectionRef and docs to write and suppose this is wrapped within an async function

var currentBatch = firestore.batch();

for ( let i = 0; i < docs.length; i++ ) {
    
    // Do some writing operation
    currentBatch.set( collectionRef.doc( i ), docs[ i ] );

    // Check current number of writes
    if ( currentBatch._writes.length === 500 ) {

        // Commit
        await currentBatch.commit();

        // Set up next batch
        currentBatch = firestore.batch();

    }

}

// Commit last batch
if ( currentBatch._writes.length !== 0 )
    await currentBatch.commit();
```

**Comment by rsgowman** [comment]
These seems reasonable, but of lower priority, so we're unlikely to be able to get to it in the near term. However, we'd welcome a PR that adds this. If this is something you'd like to tackle, we're happy to work with you on it. (Adding a `writes` property on the WriteBatch object might seem reasonable.) One warning: because this touches the public api surface, we'd have to go through our internal API Council review. As such, it could take a bit longer to reach production that you might otherwise expect.

But before you go to any effort, it's worth taking a look at your use case. The primary benefit of batches is to add the changes atomically to the Firestore database. However, your code that you provided seems willing to break apart the batches at fairly arbitrary points. (Though that may have been just as a way of an example?)

If you're attempting to write more efficiently, it would be worth benchmarking against individual writes. In many use cases, this will be just as efficient, if not possibly more so, then batches. (This differs from traditional databases where each individual write triggers an fsync() but transactions only trigger one during the commit stage.)

Another alternative that we've seen before is to simply count in the client code. However, as you've perhaps already noticed, a single document change could potentially result in multiple writes. But limiting your batch size to 250 document changes should be reasonably safe.

**Comment by jaschaio** [comment]
Interesting. The use case is indeed performance related. I have to write up to 50.000 fairly large documents as fast as possible from within a cloud function. I did not test batches vs. single writes but just supposed that single write requests are less performant then batches, especially as the documentation says “batching operations together reduces connection overhead resulting in faster data migration”.

Using an internal counter works in some cases but just as you said, there are others where an arbitrary amount of writes can happen per loop and it just seems unnecessary complicated to keep track of the counter then the batch itself.

You are right that this is not a high priority. I will do some benchmarking as suggested and might take a try at a PR.

**Comment by wilhuff** [comment]
Note that 50k sequential writes where you await each one before starting the next will perform badly and isn't what we're suggesting.

The single write equivalent of this would be to keep a rolling queue of promises and only await on the head of the queue's result once you have the max pending writes outstanding.

**Comment by jaschaio** [comment]
@wilhuff thanks for clarifying.

I gave it a try but wasn't able to make it work.

The Firebase Documentation says that the max pending writes are 500 per second, which means I should have at maximum 500 pending promises.

I did some experiments with existing promises queue libraries but either had very similar performance to using `batch` or run into `DEADLINE_EXCEEDED` errors.

I think this quite an interesting and relevant question. I noticed that the documentation was just updated on the 26th of Feburary to actually mention `batched writes are not guaranteed to be faster than a series of individual write operations`.

Do you have some dummy code on how one would keep a rolling queue of promises, checking if the max pending writes are reached and waiting for the head of the queue before adding new writes to it?

**Comment by mikelehen** [comment]
@jaschaio Can you quick confirm that you're using the "firebase-admin" or "@google-cloud/firestore" npm module for this testing?  This github repo is actually for the "firebase" npm module (which is the "mobile" SDK as opposed to the "server" SDK).  The server SDK actually lives over in https://github.com/googleapis/nodejs-firestore, and I just want to make sure we're on the same page about which SDK this is.

Anyway, A simple way of doing a rolling queue of promises (to have 25 outstanding at a time) would be something like:

```
function doWrites(collectionRef, docs) {
  let outstandingWrites = 0;
  let nextWrite = 0;
  let doneWritingResolver, doneWritingRejecter;
  let donePromise = new Promise((resolve, reject) => {
    doneWritingResolver = resolve;
    doneWritingRejecter = reject;
  });
  function sendNextWrite() {
    while (outstandingWrites < 25 && nextWrite < docs.length) {
      outstandingWrites++;
      console.log('Writing doc', nextWrite);
      collectionRef.add(docs[nextWrite++]).then(() => {
        console.log('Doc written');
        outstandingWrites--;
        sendNextWrite();
      }, doneWritingRejecter);
    }
    if (nextWrite === docs.length) {
      doneWritingResolver();
    }
  }
  sendNextWrite();
  return donePromise;
}

let docs = [];
for(let i = 0; i < 200; i++) {
  docs.push({i: i});
}
doWrites(firestore.collection('foo'), docs).then(() => console.log('All done!')).catch(console.error);
```

If you want to avoid rolling this yourself, you could try using a [semaphore](https://www.npmjs.com/package/semaphore) with the count being the maximum number of outstanding writes. The function you pass to `take` should perform the next write.

I'm not aware of any "max pending writes" limit. The 500 writes per second limit that I'm aware of [in our docs](https://firebase.google.com/docs/firestore/quotas) is:

> Maximum write rate to a collection in which documents contain sequential values in an indexed field

Firestore will push back if you attempt to write too fast.

Also just to double-check, in the example code you pasted above, you had `collectionRef.set(...)` but presumably you meant `collectionRef.add(...)` since `collectionRef.set(...)` would only be valid if `collectionRef` was actually a `DocumentReference` but in that case, you'd be overwriting the same document over and over which will run into different Firestore rate limits.
